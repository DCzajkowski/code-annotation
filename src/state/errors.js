export const initialState = [];

export const ADD = 'ca/errors/ADD';
export const REMOVE = 'ca/errors/REMOVE';

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD:
      return [...state, action.error];
    case REMOVE:
      return [
        ...state.slice(0, action.index),
        ...state.slice(action.index + 1),
      ];
    default:
      return state;
  }
};

export const add = error => dispatch => {
  console.error(error);
  if (Array.isArray(error)) {
    error.forEach(e =>
      dispatch({
        type: ADD,
        error: e,
      })
    );
    return;
  }
  dispatch({
    type: ADD,
    error,
  });
};

export const remove = index => ({
  type: REMOVE,
  index,
});

export default reducer;
