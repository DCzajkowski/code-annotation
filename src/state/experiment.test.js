import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import reducer, {
  initialState,
  LOAD,
  LOAD_SUCCESS,
  LOAD_ERROR,
  SET_EXPERIMENT,
  load,
} from './experiment';
import { ADD as ERROR_ADD } from './errors';

const mockStore = configureMockStore([thunk]);

describe('experiment/reducer', () => {
  it('LOAD', () => {
    expect(reducer(initialState, { type: LOAD })).toMatchSnapshot();
  });

  it('LOAD_SUCCESS', () => {
    expect(
      reducer(
        { ...initialState, loading: true },
        {
          type: LOAD_SUCCESS,
          index: 1,
        }
      )
    ).toMatchSnapshot();
  });

  it('LOAD_ERROR', () => {
    expect(
      reducer(
        { ...initialState, loading: true },
        { type: LOAD_ERROR, error: 'test error' }
      )
    ).toMatchSnapshot();
  });

  it('SET_EXPERIMENT', () => {
    expect(
      reducer(initialState, { type: SET_EXPERIMENT, id: 1, name: 'test name' })
    ).toMatchSnapshot();
  });
});

describe('experiment/actions', () => {
  describe('load', () => {
    it('success', () => {
      const store = mockStore({
        experiment: initialState,
      });

      fetch.mockResponse(
        JSON.stringify({
          data: {
            id: 1,
            name: 'experiment name',
          },
        })
      );

      store.dispatch(load(1)).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: LOAD,
          },
          {
            type: 'ca/experiment/SET_EXPERIMENT',
            id: 1,
            name: 'experiment name',
          },
          {
            type: 'ca/experiment/LOAD_SUCCESS',
          },
        ]);
      });
    });

    it('error', () => {
      const store = mockStore({
        experiment: initialState,
      });

      const errText = 'some error';
      fetch.mockReject(errText);

      store.dispatch(load(1)).then(() => {
        expect(store.getActions()).toEqual([
          {
            type: LOAD,
          },
          {
            type: ERROR_ADD,
            error: errText,
          },
          {
            type: LOAD_ERROR,
            error: [errText],
          },
        ]);
      });
    });
  });
});
