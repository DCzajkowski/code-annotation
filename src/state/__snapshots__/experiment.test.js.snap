// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`experiment/reducer LOAD 1`] = `
Object {
  "error": null,
  "id": null,
  "loading": true,
  "name": null,
}
`;

exports[`experiment/reducer LOAD_ERROR 1`] = `
Object {
  "error": "test error",
  "fileLoading": false,
  "id": null,
  "loading": false,
  "name": null,
}
`;

exports[`experiment/reducer LOAD_SUCCESS 1`] = `
Object {
  "error": null,
  "id": null,
  "loading": false,
  "name": null,
}
`;

exports[`experiment/reducer SET_EXPERIMENT 1`] = `
Object {
  "error": null,
  "id": 1,
  "loading": true,
  "name": "test name",
}
`;
